ml-workflow-demo-kfp-katib-mlflow/README.md

```diff
@@ -0,0 +1,472 @@
+    df_clean.loc[:,'CHURN']=pd.get_dummies(df_data['CHURN']).iloc[:, 0]
+
+    # Get rid of non-numeric columns.
+    df_clean = df_clean.select_dtypes(exclude='object')
+
+    # Save results to S3
+    csv_buffer = StringIO()
+    df_clean.to_csv(csv_buffer)
+    s3_resource = boto3.resource(
+        's3',
+        endpoint_url='http://minio.kubeflow.svc.cluster.local:9000',
+        aws_access_key_id=os.getenv('AWS_ACCESS_KEY_ID'),
+        aws_secret_access_key=os.getenv('AWS_SECRET_ACCESS_KEY')
+    )
+    check_bucket = s3_resource.Bucket(bucket)
+    if not check_bucket.creation_date:
+        # bucket does not exist, create it
+        s3_resource.create_bucket(Bucket=bucket)
+    print(f"Saving CSV of shape {df_clean.shape} to s3")
+    s3_resource.Object(bucket, key).put(Body=csv_buffer.getvalue())
+
+    return "Done"
+```
+
+8. Create the next pipeline step that will do hyperparameter tuning using Katib and a training container image `docker.io/misohu/kubeflow-training:latest`. For more details on the training container image refer to [resources README](./resources/README.md) of this guide.
+
+
+```python
+# Katib hyperparameter tuning operation.
+def create_katib_experiment_op(experiment_name, experiment_namespace, bucket, key):
+    # Trial count specification.
+    max_trial_count = 5
+    max_failed_trial_count = 3
+    parallel_trial_count = 2
+
+    # Objective specification.
+    objective = V1beta1ObjectiveSpec(
+        type="maximize",
+        goal=0.95,
+        objective_metric_name="accuracy"
+    )
+
+    # Algorithm specification.
+    algorithm = V1beta1AlgorithmSpec(
+        algorithm_name="random",
+    )
+
+    # Experiment search space.
+    # In this example we tune the number of epochs.
+    parameters = [
+        V1beta1ParameterSpec(
+            name="epochs",
+            parameter_type="int",
+            feasible_space=V1beta1FeasibleSpace(
+                min="5",
+                max="10"
+            ),
+        )
+    ]
+
+    # Experiment trial template.
+    trial_spec = {
+        "apiVersion": "kubeflow.org/v1",
+        "kind": "TFJob",
+        "spec": {
+            "tfReplicaSpecs": {
+                "Chief": {
+                    "replicas": 1,
+                    "restartPolicy": "OnFailure",
+                    "template": {
+                        "metadata": {
+                            "annotations": {
+                                "sidecar.istio.io/inject": "false"
+                            }
+                        },
+                        "spec": {
+                            "containers": [
+                                {
+                                    "name": "tensorflow",
+                                    "image": "docker.io/misohu/kubeflow-training:latest",
+                                    "command": [
+                                        "python",
+                                        "/opt/model.py",
```

Comment:
The demo really takes a while.  I think we can make some tweaks in model.py (reduce max iterations, simpler model, etc) to make this faster

Reply:
Sorry, I made this comment before I realised my run was broken and stuck.  So I'll amend it to "**if** the demo takes more than ~10 minutes, we should make some tweaks to speed it up"